File: scripts/sources.py
#!/usr/bin/env python3
"""Where postings come from.

A source obtains raw postings and normalizes them. It judges nothing: no
filtering, no database, no scoring. Every source returns the same shape, so
`ingest.py` never learns where a row came from and every filter applies to
every source.

Adding a mechanism means adding a function here and one line in REGISTRY.
Nothing downstream changes.

`rank` is source precedence, not quality: 0 sources are authoritative for a
company (the employer's own board), 1 sources are discovery (an aggregator that
may re-list what a rank-0 source already covers). Ingest resolves overlaps with
it instead of naming any source in a condition.

Every source returns `models.Posting`, so a payload that changes shape fails
here, naming the field, rather than somewhere downstream.
"""

import html
import json
import re
import time
import urllib.request

from jobkit import MAX_DESCRIPTION_CHARS, to_iso
from models import Posting

USER_AGENT = "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) job-scan/1.0"
TAG_RE = re.compile(r"<[^>]+>")
WS_RE = re.compile(r"[ \t\r\f\v]+")
BLANKS_RE = re.compile(r"\n{3,}")
VIEWJOB = "https://www.indeed.com/viewjob?jk={}"
APPLYSTART = "https://www.indeed.com/applystart?jk={}&from=vj"


def http_get_json(url, timeout=25, attempts=3):
    last_error = None
    for attempt in range(attempts):
        try:
            request = urllib.request.Request(url, headers={"User-Agent": USER_AGENT})
            with urllib.request.urlopen(request, timeout=timeout) as response:
                return json.loads(response.read().decode("utf-8", "replace"))
        except Exception as error:
            last_error = error
            if attempt < attempts - 1:
                time.sleep(1.5 * (attempt + 1))
    raise last_error


def html_to_text(raw):
    """Greenhouse ships HTML-escaped HTML, so this unescapes on both sides of the strip."""
    if not raw:
        return ""
    text = html.unescape(str(raw))
    text = re.sub(r"(?i)<br\s*/?>", "\n", text)
    text = re.sub(r"(?i)</(p|div|li|h[1-6]|tr)>", "\n", text)
    text = re.sub(r"(?i)<li[^>]*>", "- ", text)
    text = TAG_RE.sub("", text)
    text = html.unescape(text)
    text = text.replace("\xa0", " ")
    text = WS_RE.sub(" ", text)
    text = "\n".join(line.strip() for line in text.split("\n"))
    return BLANKS_RE.sub("\n\n", text).strip()


# --- Employer boards: public JSON APIs, one request per board ---------------

def greenhouse(company):
    slug = company["slug"]
    payload = http_get_json(f"https://boards-api.greenhouse.io/v1/boards/{slug}/jobs?content=true")
    out = []
    for job in payload.get("jobs", []):
        location = ((job.get("location") or {}).get("name") or "").strip()
        out.append(Posting(
            key=f"greenhouse:{slug}:{job.get('id')}",
            source="greenhouse", ats="greenhouse",
            company=company["name"], title=(job.get("title") or "").strip(),
            url=job.get("absolute_url"), apply_url=job.get("absolute_url"),
            location=location, remote=int("remote" in location.lower()),
            posted_at=to_iso(job.get("first_published") or job.get("updated_at")),
            description=html_to_text(job.get("content"))[:MAX_DESCRIPTION_CHARS]))
    return out


def _lever_description(job):
    """Lever splits a posting across descriptionPlain, the `lists` sections and
    `additional`. Requirements usually live in `lists`, so all three matter."""
    parts = [job.get("descriptionPlain") or html_to_text(job.get("description"))]
    for section in job.get("lists") or []:
        heading = (section.get("text") or "").strip()
        body = html_to_text(section.get("content"))
        if heading:
            parts.append(heading)
        if body:
            parts.append(body)
    parts.append(job.get("additionalPlain") or html_to_text(job.get("additional")))
    return "\n\n".join(part for part in parts if part)[:MAX_DESCRIPTION_CHARS]


def lever(company):
    slug = company["slug"]
    payload = http_get_json(f"https://api.lever.co/v0/postings/{slug}?mode=json")
    if not isinstance(payload, list):
        raise ValueError("lever board returned no posting list")
    out = []
    for job in payload:
        categories = job.get("categories") or {}
        locations = categories.get("allLocations") or []
        location = categories.get("location") or (locations[0] if locations else "")
        workplace = (job.get("workplaceType") or "").lower()
        salary = job.get("salaryRange") or {}
        out.append(Posting(
            key=f"lever:{slug}:{job.get('id')}",
            source="lever", ats="lever",
            company=company["name"], title=(job.get("text") or "").strip(),
            url=job.get("hostedUrl"), apply_url=job.get("applyUrl") or job.get("hostedUrl"),
            location=", ".join(locations) if locations else str(location).strip(),
            remote=int(workplace == "remote" or "remote" in str(location).lower()),
            compensation=(
                f"{salary.get('min')}-{salary.get('max')} {salary.get('currency') or ''}".strip()
                if salary.get("min") else None),
            comp_min=salary.get("min"), comp_max=salary.get("max"),
            comp_period="YEARLY" if salary.get("interval") in (None, "per-year-salary") else None,
            posted_at=to_iso(job.get("createdAt")),
            description=_lever_description(job)))
    return out


def ashby(company):
    slug = company["slug"]
    payload = http_get_json(
        f"https://api.ashbyhq.com/posting-api/job-board/{slug}?includeCompensation=true")
    out = []
    for job in payload.get("jobs", []):
        secondary = job.get("secondaryLocations") or []
        extra = [s.get("location") for s in secondary if isinstance(s, dict) and s.get("location")]
        location = ", ".join([job.get("location") or ""] + extra).strip(", ")
        compensation = job.get("compensation") or {}
        summary = compensation.get("compensationTierSummary") if isinstance(compensation, dict) else None
        out.append(Posting(
            key=f"ashby:{slug}:{job.get('id')}",
            source="ashby", ats="ashby",
            company=company["name"], title=(job.get("title") or "").strip(),
            url=job.get("jobUrl"), apply_url=job.get("applyUrl") or job.get("jobUrl"),
            location=location,
            remote=int(bool(job.get("isRemote")) or "remote" in location.lower()),
            compensation=summary,
            expired=int(job.get("isListed") is False),
            posted_at=to_iso(job.get("publishedAt")),
            description=(job.get("descriptionPlain")
                         or html_to_text(job.get("descriptionHtml")))[:MAX_DESCRIPTION_CHARS]))
    return out


# --- Aggregators: a browser fetches, this reads what it saved ---------------

def indeed(harvest_path):
    """Indeed serves navigation without complaint and throttles fetch()/XHR
    against the same URLs, so a browser does the fetching and this only parses
    the harvest it saved. Descriptions are fetched later, for kept rows only."""
    with open(harvest_path, "r", encoding="utf-8") as handle:
        payload = json.load(handle)
    if isinstance(payload, dict):
        blocks = payload.get("results") or []
        cards = [c for b in blocks for c in (b.get("rows") or [])] if blocks else payload.get("rows", [])
    else:
        cards = payload

    out = []
    for card in cards:
        jobkey = card.get("jobkey")
        if not jobkey:
            continue
        salary = card.get("extractedSalary") or {}
        remote = (card.get("remoteWorkModel") or {}).get("type") or ""
        location = card.get("formattedLocation") or ""
        low, high = salary.get("min"), salary.get("max")
        stated = None
        if low or high:
            unit = (salary.get("type") or "").lower()
            stated = (f"{low:,.0f}-{high:,.0f} {unit}".strip() if low and high
                      else f"{(low or high):,.0f} {unit}".strip())
        out.append(Posting(
            key=f"indeed:{jobkey}",
            source="indeed", ats=None,
            company=card.get("company") or "", title=(card.get("title") or "").strip(),
            url=VIEWJOB.format(jobkey), apply_url=APPLYSTART.format(jobkey),
            location=location,
            remote=int(bool(remote) or "remote" in location.lower()),
            compensation=stated,
            comp_min=low, comp_max=high,
            comp_period=(salary.get("type") or "").upper() or None,
            sponsored=int(bool(card.get("sponsored"))),
            expired=int(bool(card.get("expired"))),
            posted_at=to_iso(card.get("pubDate")),
            raw=json.dumps(card, ensure_ascii=False)))
    return out


# --- The registry. Adding a mechanism is one line. -------------------------

REGISTRY = {
    "greenhouse": {"fetch": greenhouse, "kind": "board", "rank": 0},
    "lever":      {"fetch": lever,      "kind": "board", "rank": 0},
    "ashby":      {"fetch": ashby,      "kind": "board", "rank": 0},
    "indeed":     {"fetch": indeed,     "kind": "harvest", "rank": 1},
}

BOARDS = {name: s["fetch"] for name, s in REGISTRY.items() if s["kind"] == "board"}
RANK = {name: s["rank"] for name, s in REGISTRY.items()}

File: lib/sources.ts
import path from "node:path";

import { ROOT } from "./root.ts";
import { MAX_DESCRIPTION_CHARS, htmlToText, toIso } from "./text.ts";
import { posting, type Posting } from "./posting.ts";

const USER_AGENT = "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) job-scan/1.0";
const VIEWJOB = "https://www.indeed.com/viewjob?jk=";
const APPLYSTART = "https://www.indeed.com/applystart?jk=";

export type Company = { name: string; ats: string; slug: string };
type Json = Record<string, any>;

const sleep = (ms: number) => new Promise((done) => setTimeout(done, ms));

export async function getJson(url: string, timeout = 25000, attempts = 3): Promise<any> {
  let last: unknown;
  for (let attempt = 0; attempt < attempts; attempt++) {
    try {
      const answered = await fetch(url, {
        headers: { "User-Agent": USER_AGENT },
        signal: AbortSignal.timeout(timeout),
      });
      if (!answered.ok) throw new Error(`HTTP ${answered.status} ${answered.statusText}`);
      return await answered.json();
    } catch (error) {
      last = error;
      if (attempt < attempts - 1) await sleep(1500 * (attempt + 1));
    }
  }
  throw last;
}

export const ENDPOINT: Record<string, string> = {
  greenhouse: "https://boards-api.greenhouse.io/v1/boards/{slug}/jobs?content=true",
  lever: "https://api.lever.co/v0/postings/{slug}?mode=json",
  ashby: "https://api.ashbyhq.com/posting-api/job-board/{slug}?includeCompensation=true",
  indeed: "https://api.apify.com/v2/acts/misceres~indeed-scraper/run-sync-get-dataset-items",
};

export const QUIRK: Record<string, string> = {
  greenhouse:
    "`content` is HTML-escaped HTML, unescaped twice here. Without ?content=true there are no " +
    "descriptions to score on",
  lever:
    "returns a BARE ARRAY; createdAt is epoch ms; a posting splits across descriptionPlain, " +
    "`lists` and `additional`. A live board with no postings returns [] at 200, a wrong slug 404s",
  ashby:
    "best payload: descriptionPlain needs no HTML handling, isRemote is a real boolean, " +
    "compensation bands come back. isListed:false normalizes to expired",
  indeed:
    "a paid Apify actor, billed per listing, so `--max` is a budget and not a page size. " +
    "Descriptions arrive with the search, so there is no second pass. It never states `sponsored`, " +
    "so that filter cannot rule on an Indeed row",
};

const clipped = (text: string) => text.slice(0, MAX_DESCRIPTION_CHARS);

async function greenhouse(company: Company): Promise<Posting[]> {
  const { slug } = company;
  const payload: Json = await getJson(ENDPOINT.greenhouse.replace("{slug}", slug));
  return (payload.jobs ?? []).map((job: Json) => {
    const location = String(job.location?.name ?? "").trim();
    return posting({
      key: `greenhouse:${slug}:${job.id}`,
      source: "greenhouse",
      ats: "greenhouse",
      company: company.name,
      title: job.title,
      url: job.absolute_url,
      apply_url: job.absolute_url,
      location,
      remote: location.toLowerCase().includes("remote"),
      posted_at: toIso(job.first_published ?? job.updated_at),
      description: clipped(htmlToText(job.content)),
    });
  });
}

const leverDescription = (job: Json) => {
  const parts = [job.descriptionPlain || htmlToText(job.description)];
  for (const section of job.lists ?? []) {
    const heading = String(section.text ?? "").trim();
    const body = htmlToText(section.content);
    if (heading) parts.push(heading);
    if (body) parts.push(body);
  }
  parts.push(job.additionalPlain || htmlToText(job.additional));
  return clipped(parts.filter(Boolean).join("\n\n"));
};

async function lever(company: Company): Promise<Posting[]> {
  const { slug } = company;
  const payload = await getJson(ENDPOINT.lever.replace("{slug}", slug));
  if (!Array.isArray(payload)) throw new Error("lever board returned no posting list");
  return payload.map((job: Json) => {
    const categories = job.categories ?? {};
    const locations: string[] = categories.allLocations ?? [];
    const location = categories.location ?? locations[0] ?? "";
    const workplace = String(job.workplaceType ?? "").toLowerCase();
    const salary = job.salaryRange ?? {};
    return posting({
      key: `lever:${slug}:${job.id}`,
      source: "lever",
      ats: "lever",
      company: company.name,
      title: job.text,
      url: job.hostedUrl,
      apply_url: job.applyUrl || job.hostedUrl,
      location: locations.length ? locations.join(", ") : String(location).trim(),
      remote: workplace === "remote" || String(location).toLowerCase().includes("remote"),
      compensation: salary.min
        ? `${salary.min}-${salary.max} ${salary.currency ?? ""}`.trim()
        : null,
      comp_min: salary.min,
      comp_max: salary.max,
      comp_period:
        salary.interval == null || salary.interval === "per-year-salary" ? "YEARLY" : null,
      posted_at: toIso(job.createdAt),
      description: leverDescription(job),
    });
  });
}

async function ashby(company: Company): Promise<Posting[]> {
  const { slug } = company;
  const payload: Json = await getJson(ENDPOINT.ashby.replace("{slug}", slug));
  return (payload.jobs ?? []).map((job: Json) => {
    const extra = (job.secondaryLocations ?? [])
      .map((held: Json) => held?.location)
      .filter(Boolean);
    const location = [job.location ?? "", ...extra].join(", ").replace(/^[,\s]+|[,\s]+$/g, "");
    const compensation = job.compensation ?? {};
    return posting({
      key: `ashby:${slug}:${job.id}`,
      source: "ashby",
      ats: "ashby",
      company: company.name,
      title: job.title,
      url: job.jobUrl,
      apply_url: job.applyUrl || job.jobUrl,
      location,
      remote: Boolean(job.isRemote) || location.toLowerCase().includes("remote"),
      compensation: compensation.compensationTierSummary ?? null,
      expired: job.isListed === false,
      posted_at: toIso(job.publishedAt),
      description: clipped(job.descriptionPlain || htmlToText(job.descriptionHtml)),
    });
  });
}

export type Search = { query: string; location: string; country: string; max: number };

const APIFY_ACTOR = "misceres~indeed-scraper";
const APIFY_RUN = `https://api.apify.com/v2/acts/${APIFY_ACTOR}/run-sync-get-dataset-items`;

function apifyToken(): string {
  const held = process.env.APIFY_TOKEN?.trim();
  if (!held)
    throw new Error(
      "APIFY_TOKEN is not set. Get a token from apify.com/settings/integrations and put " +
      `APIFY_TOKEN=… in ${path.join(ROOT, ".env.local")}; job-scan boards runs without one`);
  return held;
}

const PERIOD: Record<string, string> = {
  hour: "HOURLY", day: "DAILY", week: "WEEKLY", month: "MONTHLY", year: "YEARLY",
};

function statedPay(salary: string | null) {
  const said = String(salary ?? "");
  const unit = /\b(hour|day|week|month|year)/i.exec(said);
  const period = unit ? PERIOD[unit[1].toLowerCase()] : null;
  const amounts = [...said.matchAll(/([\d,]+(?:\.\d+)?)/g)]
    .map((found) => Number(found[1].replace(/,/g, "")))
    .filter((amount) => Number.isFinite(amount) && amount > 0);
  if (!amounts.length) return { min: null, max: null, period };
  const low = Math.min(...amounts);
  const high = Math.max(...amounts);
  if (/\bup to\b/i.test(said)) return { min: null, max: high, period };
  if (/\bfrom\b|\bstarting at\b/i.test(said)) return { min: low, max: null, period };
  return { min: low, max: high, period };
}

const SPAN: Record<string, number> = {
  hour: 3600e3, day: 86400e3, week: 7 * 86400e3, month: 30 * 86400e3,
};

function postedIso(item: Json): string | null {
  const dated = toIso(item.postingDateParsed ?? item.datePublished ?? null);
  if (dated) return dated;
  const said = String(item.postedAt ?? "").toLowerCase();
  if (!said) return null;
  if (/just posted|today/.test(said)) return toIso(new Date().toISOString());
  const ago = /(\d+)\+?\s*(hour|day|week|month)/.exec(said);
  if (!ago) return null;
  return toIso(new Date(Date.now() - Number(ago[1]) * SPAN[ago[2]]).toISOString());
}

export function fromApify(items: Json[]): Posting[] {
  return items
    .filter((item) => item.id && !item.error)
    .map((item) => {
      const location = String(item.location ?? "").trim();
      const title = String(item.positionName ?? "");
      const pay = statedPay(item.salary ?? null);
      const written = String(item.description ?? "").trim()
        || htmlToText(String(item.descriptionHTML ?? ""));
      return posting({
        key: `indeed:${item.id}`,
        source: "indeed",
        ats: null,
        company: item.company ?? "",
        title,
        url: item.url ?? VIEWJOB + item.id,
        apply_url: item.externalApplyLink ?? `${APPLYSTART}${item.id}&from=vj`,
        location,
        remote: /remote/i.test(location) || /remote/i.test(title),
        compensation: item.salary ?? null,
        comp_min: pay.min,
        comp_max: pay.max,
        comp_period: pay.period,
        expired: Boolean(item.isExpired),
        posted_at: postedIso(item),
        description: written ? clipped(written) : null,
        raw: JSON.stringify(item),
      });
    });
}

async function indeed(search: Search): Promise<Posting[]> {
  const token = apifyToken();
  const answered = await fetch(APIFY_RUN, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${token}`,
      "User-Agent": USER_AGENT,
    },
    body: JSON.stringify({
      position: search.query,
      location: search.location,
      country: search.country,
      maxItemsPerSearch: search.max,
      parseCompanyDetails: false,
      saveOnlyUniqueItems: true,
    }),
    signal: AbortSignal.timeout(300000),
  });
  if (!answered.ok)
    throw new Error(
      `HTTP ${answered.status} ${answered.statusText}: ${(await answered.text()).slice(0, 200)}`);
  const items = await answered.json();
  return fromApify(Array.isArray(items) ? items : []);
}

type Entry =
  | { fetch: (company: Company) => Promise<Posting[]>; kind: "board"; rank: number }
  | { fetch: (search: Search) => Promise<Posting[]>; kind: "search"; rank: number };

export const REGISTRY: Record<string, Entry> = {
  greenhouse: { fetch: greenhouse, kind: "board", rank: 0 },
  lever: { fetch: lever, kind: "board", rank: 0 },
  ashby: { fetch: ashby, kind: "board", rank: 0 },
  indeed: { fetch: indeed, kind: "search", rank: 1 },
};

const listed = (held: Record<string, unknown>) => Object.keys(held).sort().join(",");
if (listed(REGISTRY) !== listed(ENDPOINT) || listed(REGISTRY) !== listed(QUIRK))
  throw new Error("a source is missing an endpoint or a quirk");

export const BOARDS = Object.fromEntries(
  Object.entries(REGISTRY)
    .filter(([, entry]) => entry.kind === "board")
    .map(([name, entry]) => [name, entry.fetch as (company: Company) => Promise<Posting[]>]),
);

const RANK = Object.fromEntries(Object.entries(REGISTRY).map(([name, e]) => [name, e.rank]));

export const rank = (source: string | null | undefined) => RANK[source ?? ""] ?? 99;

export function describe() {
  const entries = Object.entries(REGISTRY).sort(
    ([a, one], [b, two]) => one.rank - two.rank || a.localeCompare(b),
  );
  for (const [name, entry] of entries) {
    console.log(`${name}  [${entry.kind}, rank ${entry.rank}]`);
    console.log(`  ${ENDPOINT[name]}`);
    console.log(`  ${QUIRK[name]}\n`);
  }
}
